src/crypto/tls/ech.go

```diff
@@ -0,0 +1,824 @@
+	echHpkeInfoConfigId           = "tls13 ech config id"
+	echHpkeInfoSetup              = "tls13 ech"
+	echHpkeInfoSetupHrr           = "tls13 ech hrr"
+	echHpkeHrrKeyExportechContext = "tls13 ech hrr key"
+	echHpkeHrrKeyId               = "hrr key"
+	echHpkeHrrKeyLen              = 16
+
+	// Stand-in values for algorithms that are unknown prior to a particular
+	// operation.
+	//
+	// There is a slight mismatch between the API exported by HPKE and the API
+	// required to implement the ECH logic in the TLS stack. Namely, an "HPKE
+	// ciphersuite" is a (KEM, KDF, AEAD) triple, and as such, the API of the
+	// HPKE implementation we use, github.com/cisco/go-hpke, presumes all three
+	// algorithms are known prior to any HPKE operation. In contrast, an "ECH
+	// ciphersuite" is a (KDF, AEAD) pair, meaning the KDF and AEAD may not be
+	// known when doing a KEM operation, e.g., generating a KEM key pair. This
+	// library provides a thin wrapper around github.com/cisco/go-hpke that
+	// resolves this mismatch.
+	dummyKemId  = uint16(hpke.DHKEM_X25519)
+	dummyKdfId  = uint16(hpke.KDF_HKDF_SHA256)
+	dummyAeadId = uint16(hpke.AEAD_AESGCM128)
+
+	// The ciphertext expansion incurred by the AEAD identified by dummyAeadId.
+	dummyAeadOverheadLen = 16
+
+	// The output size of "Expand()" for the KDF identified by dummyKdfId.
+	dummyKdfOutputLen = 32
+
+	// The maximum output length of the Extract() operation among the set of
+	// supported KDFs. Currently this is 64, which is the number of extracted
+	// bytes for HKDF-SHA512.
+	maxHpkeKdfExtractLen uint16 = 64
+)
+
+// The public used to generate covertext for the ECH extension. This is a random
+// 32-byte string, which will be interpreted as an X25519 public key.
+var echDummyPublicKey = []byte{
+	143, 38, 37, 36, 12, 6, 229, 30, 140, 27, 167, 73, 26, 100, 203, 107, 216,
+	81, 163, 222, 52, 211, 54, 210, 46, 37, 78, 216, 157, 97, 241, 244,
+}
+
+func (c *Conn) echOfferOrBypass(helloBase *clientHelloMsg) (hello, helloInner *clientHelloMsg, err error) {
+	config := c.config
+
+	// Trigger failure scenarios, if testing.
+	echConfigs := config.ClientECHConfigs
+	if c.hrrTriggered && testingTriggerECHBypassAfterHRR {
+		echConfigs = nil
+	}
+	if !c.hrrTriggered && testingTriggerECHBypassBeforeHRR {
+		echConfigs = nil
+	}
+
+	// Select an ECH configuration suitable for this connection.
+	var echConfig *ECHConfig
+	for _, conf := range echConfigs {
+		// A suitable configuration is one that offers an HPKE ciphersuite that
+		// is supported by the client and indicates the version of ECH
+		// implemented by this TLS client.
+		if conf.isSupported() && extensionECH == conf.version {
+			echConfig = &conf
+			break
+		}
+	}
+
+	hello = new(clientHelloMsg)
+	*hello = *helloBase
+
+	// Decide whether to offer the ECH extension in this connection, If offered,
+	// then hello is set to the ClientHelloOuter and helloInner is set to the
+	// ClientHelloInner.
+	if echConfig != nil &&
+		config.ECHEnabled &&
```

Comment:
Perhaps `ECHEnabled` could be checked much earlier (e.g. the begin of the function) and avoid doing logic when not enabled? Likewise for the `if` case after this one.